### PR TITLE
RFC-0027: PyTorch DistributedTensor

### PR DESCRIPTION
**Pull request by wanchaol**
This PR proposes RFC: PyTorch DistributedTensor

### CODE REVIEW

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
+# RFC: PyTorch DistributedTensor
+
+We proposed distributed tensor primitives to allow easier distributed computation authoring in SPMD manner, the primitives are simple but powerful to express tensor distributions with both sharding and replication strategies. This could empower native PyTorch Tensor Parallelism and advanced parallelism explorations. For example, to shard a big tensor across devices with 3 lines of code:
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:3]:
```suggestion
We propose distributed tensor primitives to allow easier distributed computation authoring in SPMD(Single Program Multiple Devices) paradigm. The primitives are simple but powerful when used to express tensor distributions with both sharding and replication parallelism strategies. This could empower native Tensor parallelism among other advanced parallelism explorations. For example, to shard a big tensor across devices with 3 lines of code:
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+```python
+import torch  
+from torch.distributed import DeviceMesh, Shard, distribute_tensor  
+  
+mesh = DeviceMesh("cuda", list(range(world_size)))
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:9]:
```suggestion
# Create a mesh topology with the available devices.
mesh = DeviceMesh("cuda", list(range(world_size)))  
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+We proposed distributed tensor primitives to allow easier distributed computation authoring in SPMD manner, the primitives are simple but powerful to express tensor distributions with both sharding and replication strategies. This could empower native PyTorch Tensor Parallelism and advanced parallelism explorations. For example, to shard a big tensor across devices with 3 lines of code:
+
+```python
+import torch  
+from torch.distributed import DeviceMesh, Shard, distribute_tensor  
+  
+mesh = DeviceMesh("cuda", list(range(world_size)))  
+big_tensor = torch.randn(100000, 88)  
+distribute_tensor(big_tensor, mesh, [Shard(dim=0)])
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:11]:
```suggestion
# Shard this tensor over the mesh by sharding `big_tensor`'s 0th dimension over the 0th dimension of `mesh`.
distribute_tensor(big_tensor, mesh, [Shard(dim=0)])
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+```
+
+To see a complete design doc for this proposal, please refer to https://docs.google.com/document/d/1nFeJ8NSFNhNlCkNgWK31ZGRqm1L9rd0i_XN_RprphaI/edit#heading=h.6sovjqv9jiqn
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:14]:
```suggestion
To see a complete design doc for this proposal, please refer to this [doc](https://docs.google.com/document/d/1nFeJ8NSFNhNlCkNgWK31ZGRqm1L9rd0i_XN_RprphaI/edit#heading=h.6sovjqv9jiqn)
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+## Motivation
+
+Today there are mainly three ways to scale up distributed training: Data Parallel, Tensor Parallel and Pipeline Parallel. Each of them works on a separate dimension where solutions have been built independently (i.e. PyTorch DDP, FSDP, ShardedTensor, PiPPy, etc.). When training really large models, users would like to use these technologies together (i.e. 3-D Parallelism), while the interoperability of the existing solutions are not great and often hard to use (i.e. users might want arbitrary combinations of the data parallel, tensor parallel and pipeline parallel). This is becoming an issue for users and one of the biggest reasons is that there’s no common abstractions that build the bridge between different parallel strategies.
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:18]:
```suggestion
Today there are mainly three ways to scale up distributed training: Data Parallel, Tensor Parallel and Pipeline Parallel. Each of them works on a separate dimension where solutions have been built independently (i.e. PyTorch DDP, FSDP, ShardedTensor, PiPPy, etc.). When training really large models, users would like to use these technologies together (i.e. 3-D Parallelism), while the interoperability of the existing solutions are not great and often hard to use (i.e. users might want arbitrary combinations of the data parallel, tensor parallel and pipeline parallel). This is becoming an issue for users and one of the biggest reasons is that there’s no common abstractions that build the bridge between different parallelism strategies.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+An ideal scenario is that users could just build their models like in a single node/device, without worrying about how to do distributed training in a cluster, and our solutions could help them run distributed training in an efficient manner. For example, researchers just need to build their big transformer model, and PyTorch Distributed automatically figures out how to split the model and run pipeline parallel across different nodes, how to run data parallel and tensor parallel within each node. In order to achieve this, we need to translate a single device model into a distributed version and train/serve it with our runtime. To represent the distributed version of the model and facilitate translations, we need some common abstractions to represent distribution and computation.
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:20]: nit: This seems to be describing the compiler approach and we can maybe skip this according to me.

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  
+There're many recent works that working on tensor level parallelism to provide common abstractions, see the `Related Works` in the last section for more details. Inspired by [GSPMD](https://arxiv.org/pdf/2105.04663.pdf), [Oneflow](https://arxiv.org/pdf/2110.15032.pdf) and [TF’s DTensor](https://www.tensorflow.org/guide/dtensor_overview), we introduce a DistributedTensor concept to represent generic data distributions across hosts. DistributedTensor is the next evolution of ShardedTensor and provides basic abstractions to distribute storage and compute. It serves as one of the basic building blocks for distributed program translations and describes the layout of a distributed training program. With the DistributedTensor abstraction, we can build different parallelism strategies in a easy way, including generic tensor parallelism, or DDP/FSDP parallelism patterns.
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:22]:
```suggestion
There're many recent works that working on tensor level parallelism to provide common abstractions, see the `Related Works` in the last section for more details. Inspired by [GSPMD](https://arxiv.org/pdf/2105.04663.pdf), [Oneflow](https://arxiv.org/pdf/2110.15032.pdf) and [TF’s DTensor](https://www.tensorflow.org/guide/dtensor_overview), we introduce a DistributedTensor concept to represent generic data distributions across hosts. DistributedTensor is the next evolution of ShardedTensor and provides basic abstractions to distribute storage and compute. It serves as one of the basic building blocks for distributed program translations and describes the layout of a distributed training program. With the DistributedTensor abstraction, we can seamlessly build parallelism strategies such as tensor parallelism, DDP and FSDP.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+## Value Propsition
+
+The primary value of DistributedTensor includes:
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:26]:
```suggestion
DistributedTensor primarily:
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+-   Offers a uniform way to save/load state dict during checkpointing, even when there’re complex data distribution strategies
+-   DistributedTensor could natively offer Tensor Parallelism solution in eager mode, just like our current ShardedTensor solution. Moreover, it gives additional flexibility for advanced users who want to mix sharding and replication.
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:28]:
```suggestion
-   Could natively offer Tensor Parallelism solution in eager mode, just like our current ShardedTensor solution. Moreover, it gives additional flexibility for advanced users who want to mix sharding and replication.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+-   DistributedTensor could be used as a basic building block of a compiler based solution to do distributed training
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:29]:
```suggestion
-   Could be used as a basic building block of a compiler based solution to do distributed training    
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+## Value Propsition
+
+The primary value of DistributedTensor includes:
+-   Offers a uniform way to save/load state dict during checkpointing, even when there’re complex data distribution strategies
+-   DistributedTensor could natively offer Tensor Parallelism solution in eager mode, just like our current ShardedTensor solution. Moreover, it gives additional flexibility for advanced users who want to mix sharding and replication.
+-   DistributedTensor could be used as a basic building block of a compiler based solution to do distributed training    
+-   DistributedTensor could act as a SPMD programming model entry point for ML System Engineers, providing good UX to mix up different types of parallelism.
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:30]:
```suggestion
-   Could act as a SPMD programming model entry point for ML System Engineers, providing good UX to mix up different types of parallelism.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  
+There're many recent works that working on tensor level parallelism to provide common abstractions, see the `Related Works` in the last section for more details. Inspired by [GSPMD](https://arxiv.org/pdf/2105.04663.pdf), [Oneflow](https://arxiv.org/pdf/2110.15032.pdf) and [TF’s DTensor](https://www.tensorflow.org/guide/dtensor_overview), we introduce a DistributedTensor concept to represent generic data distributions across hosts. DistributedTensor is the next evolution of ShardedTensor and provides basic abstractions to distribute storage and compute. It serves as one of the basic building blocks for distributed program translations and describes the layout of a distributed training program. With the DistributedTensor abstraction, we can build different parallelism strategies in a easy way, including generic tensor parallelism, or DDP/FSDP parallelism patterns.
+
+## Value Propsition
+
+The primary value of DistributedTensor includes:
+-   Offers a uniform way to save/load state dict during checkpointing, even when there’re complex data distribution strategies
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:27]:
```suggestion
-   Offers a uniform way to save/load state dict during checkpointing, even when there’re complex data distribution strategies such as combing tensor parallelism with parameter sharding in FSDP.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md (hunk 11 above)
Comment (anj-s) [reply RFC-0027-PyTorch-Distributed-Tensor.md:30]: The last 2 points could be combined to be one point. WDYT?

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+-   DistributedTensor could natively offer Tensor Parallelism solution in eager mode, just like our current ShardedTensor solution. Moreover, it gives additional flexibility for advanced users who want to mix sharding and replication.
+-   DistributedTensor could be used as a basic building block of a compiler based solution to do distributed training    
+-   DistributedTensor could act as a SPMD programming model entry point for ML System Engineers, providing good UX to mix up different types of parallelism.
+
+## PyTorch Distributed Tensor
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:32]:
```suggestion
## PyTorch DistributedTensor
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+### DistributedTensor API
+
+We offer both lower level DistributedTensor API and module level API to create nn.Module with “distributed” parameters.
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:36]:
```suggestion
We offer both a lower level DistributedTensor API and a module level API to create a `nn.Module` with “distributed” parameters.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+#### Basic DistributedTensor API Examples
+
+Here are some basic DistributedTensor level API examples that showcase:
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:40]:
```suggestion
Here are some basic DistributedTensor API examples that showcase: 
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+## PyTorch Distributed Tensor
+
+### DistributedTensor API
+
+We offer both lower level DistributedTensor API and module level API to create nn.Module with “distributed” parameters.
+
+#### Basic DistributedTensor API Examples
+
+Here are some basic DistributedTensor level API examples that showcase: 
+1.  How to construct a DistributedTensor directly, to represent different types of sharding, replication, sharding + replication
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:41]:
```suggestion
1. How to construct a DistributedTensor directly, to represent different types of sharding, replication, sharding + replication strategies.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+2. How to create DistributedTensor from a local torch.Tensor
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:42]:
```suggestion
2. How to create DistributedTensor from a local `torch.Tensor`.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+3. How to “reshard” an existing DistributedTensor to a different DistributedTensor with different placement strategy, or world size.
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:43]:
```suggestion
3. How to “reshard” an existing DistributedTensor to a different DistributedTensor with modified placement strategy or world size.
```

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+from torch.distributed import DTensor, DeviceMesh, Shard, Replicate, distribute_module  
+  
+# construct a device mesh with available devices (multi-host or single host)  
+device_mesh = DeviceMesh(device_type="cuda", [0, 1, 2, 3])  
+# if we want to do row-wise sharding  
+rowwise_placement=[Shard(0)]  
+# if we want to do col-wise sharding  
+colwise_placement=[Shard(1)]  
+# distributed tensor returned will be sharded across the dimension specified in placements  
+distributed.empty((8, 12), device_mesh=device_mesh, placements=rowwise_placement)
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:57]: is `distributed.empty` a new API that we are proposing over `torch.empty`?

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+  
+# if we want to distributed a tensor with both replication and sharding  
+device_mesh = DeviceMesh(device_type="cuda", [[0, 1], [2, 3]])  
+# replicate across the first dimension of device mesh, then sharding on the second dimension of device mesh  
+spec=[Replicate(), Shard(0)]  
+distributed.empty((8, 8), device_mesh=device_mesh, placements=spec)  
+  
+# create a DistributedTensor that shards on dim 0, from a local torch.Tensor  
+local_tensor = torch.randn((8, 8), requires_grad=True)  
+rowwise_tensor = DTensor.from_local(local_tensor, device_mesh, rowwise_placement)
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:72]: Can we add a comment about how `from_local` is different from calling DTensor(torch.empty())?

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+        super.__init__()  
+        self.fc1 = nn.Linear(8, 8)  
+        self.fc2 = nn.Linear(8, 8)  
+        self.relu = nn.ReLU()  
+     
+    def forward(self, input):  
+        return self.relu(self.fc1(input) + self.fc2(input))  
+  
+def shard_params(mod_name, mod):  
+    def to_dist_tensor(t): return distribute_tensor(t, mesh, rowwise_placement)
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:117]: Do `mesh` and `rowwise_placement` need to be input to `shard_params`?

File: RFC-0027-PyTorch-Distributed-Tensor.md [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+-   DTensor implements MLIR passes to do propagation and operator implementations.
+
+There are also several cutting edge research fields that embeds tensor sharding as part of the system, i.e. [Megatron-LM](https://arxiv.org/pdf/1909.08053.pdf) for tensor parallelism on Transformer based models. [DeepSpeed](https://github.com/microsoft/DeepSpeed) for training large scale models with different optimization techniques on top of tensor sharding.
+
+In PyTorch, we have existing [ShardedTensor](https://docs.google.com/document/d/1WEjwKYv022rc1lSrYcNWh3Xjx9fJC7zsrTKTg0wbPj0/edit?usp=sharing) work in the prototype stage, which introduces basic PyTorch sharding primitives as our Tensor Parallelism solution. But ShardedTensor only has tensor sharding support, which makes it hard to be used by users to describe other data distributions strategies like replication or replication + sharding. As a distributed system developer who wants to explore more parallelism patterns, it’s crucial to have a basic building block that describes the data distribution in a uniform way. This DistributedTensor RFC aims at solving this and provide a fundamental abstraction for distributed training.
```
Comment (anj-s) [inline RFC-0027-PyTorch-Distributed-Tensor.md:155]: Is this doc also in the public folder?

File: RFC-0027-PyTorch-Distributed-Tensor.md (hunk 21 above)
Comment (wanchaol) [reply RFC-0027-PyTorch-Distributed-Tensor.md:119]: `partition_fn` should have a fixed param signature list, I think `mesh` and `placements` are sth user can construct inside the `shard_params` function. I'll update it accordingly.
